modules/library/main/src/main/java/org/geotools/renderer/crs/ProjectionHandler.java

```diff
@@ -343,29 +306,132 @@ private void addTransformedEnvelope(ReferencedEnvelope re, List<ReferencedEnvelo
         // have no idea at all) we don't actually split, but add elements
         List<ReferencedEnvelope> envelopes = new ArrayList<ReferencedEnvelope>();
         envelopes.add(re);
-        if (re.getMinX() < -180) {
-            envelopes.add(
-                    new ReferencedEnvelope(
-                            re.getMinX() + 360,
-                            Math.min(re.getMaxX() + 360, 180),
-                            re.getMinY(),
-                            re.getMaxY(),
-                            re.getCoordinateReferenceSystem()));
-        }
-        if (re.getMaxX() > 180) {
-            envelopes.add(
-                    new ReferencedEnvelope(
-                            Math.max(re.getMinX() - 360, -180),
-                            re.getMaxX() - 360,
-                            re.getMinY(),
-                            re.getMaxY(),
-                            re.getCoordinateReferenceSystem()));
-        }
+        adjustEnvelope(re, envelopes, false);
         mergeEnvelopes(envelopes);
         reprojectEnvelopes(sourceCRS, envelopes);
         return envelopes.stream().filter(e -> e != null).collect(Collectors.toList());
     }
 
+    /**
+     * Adjust the envelope by taking into account dateline wrapping as well as multiple spans of the
+     * whole world extent. When transform flag is true, the envelopes will be transformed before
+     * being returned
+     */
+    private void adjustEnvelope(
```

Comment:
Could you add a good set of tests inside ProjectionHandlerTest covering all the possible cases handled here?

Reply:
will do.